src/coreclr/jit/lsraxarch.cpp

```diff
@@ -2334,47 +2334,100 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
-                    srcCount += BuildDelayFreeUses(op3, op1);
+
+                    if (op2->isContained() || op2->IsRegOptional())
+                    {
+                        // op1 = (op1 * [op2]) + op3
+                        srcCount += op2->isContained() ? BuildOperandUses(op2) : BuildDelayFreeUses(op2, op1);
+                        srcCount += BuildDelayFreeUses(op3, op1);
+                    }
+                    else
+                    {
+                        // op1 = (op1 * op2) + [op3]
+                        srcCount += op3->isContained() ? BuildOperandUses(op3) : BuildDelayFreeUses(op3, op1);
+                        srcCount += BuildDelayFreeUses(op2, op1);
+                    }
                 }
-                else if (op1->isContained())
+                else if (overwrittenOpNum == 2)
                 {
-                    // 231 form: op3 = (op2 * op3) + [op1]
-
-                    tgtPrefUse = BuildUse(op3);
-
-                    srcCount += BuildOperandUses(op1);
-                    srcCount += BuildDelayFreeUses(op2, op1);
+                    tgtPrefUse = BuildUse(op2);
                     srcCount += 1;
+
+                    if (op1->isContained())
+                    {
+                        // op2 = ([op1] * op2) + op3
+                        srcCount += BuildOperandUses(op1);
+                        srcCount += BuildDelayFreeUses(op3, op2);
+                    }
+                    else
+                    {
+                        // op2 = (op1 * op2) + [op3]
+                        srcCount += BuildDelayFreeUses(op1, op2);
+                        srcCount += op3->isContained() ? BuildOperandUses(op3) : BuildDelayFreeUses(op3, op1);
+                    }
                 }
-                else
+                else if (overwrittenOpNum == 3)
                 {
-                    // 213 form: op1 = (op2 * op1) + [op3]
-
-                    tgtPrefUse = BuildUse(op1);
+                    tgtPrefUse = BuildUse(op3);
                     srcCount += 1;
 
-                    if (copiesUpperBits)
+                    if (op1->isContained())
                     {
-                        srcCount += BuildDelayFreeUses(op2, op1);
+                        // op3 = ([op1] * op2) + op3
+                        srcCount += BuildOperandUses(op1);
+                        srcCount += BuildDelayFreeUses(op2, op3);
                     }
                     else
                     {
-                        tgtPrefUse2 = BuildUse(op2);
+                        // op3 = (op1 * [op2]) + op3
+                        srcCount += op2->isContained() ? BuildOperandUses(op2) : BuildDelayFreeUses(op2, op3);
+                        srcCount += BuildDelayFreeUses(op1, op3);
+                    }
+                }
+                else
+                {
+                    assert(overwrittenOpNum == 0);
+                    if (op1->isContained())
```

Comment:
Can you add a comment why we chose to check for the containment of `op1` for cases where no operand is overwritten?